# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/mac/CachedTextMarker.mm
# Commit: 214be178df88
# Full Hash: 214be178df88eafd0afb48bff78a27e2d199a7ae
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2023-03-02 04:57:23
# Regressor Bug: 1818450
# File Overlap Count: 1
# Description:
#   Bug 1818450 - Support getting attributed text with empty inputs. r=morgan
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D170939
# ==============================================================================

diff -r 1440eb683cd1 -r 214be178df88 accessible/mac/CachedTextMarker.mm
--- a/accessible/mac/CachedTextMarker.mm	Wed Mar 01 18:58:18 2023 +0000
+++ b/accessible/mac/CachedTextMarker.mm	Wed Mar 01 19:00:08 2023 +0000
@@ -268,9 +268,13 @@
           : mRange;
 
   for (TextLeafRange segment : range) {
-    segment.Start().mAcc->AppendTextTo(
-        text, segment.Start().mOffset,
-        segment.End().mOffset - segment.Start().mOffset);
+    TextLeafPoint start = segment.Start();
+    if (start.mAcc->IsTextField() && start.mAcc->ChildCount() == 0) {
+      continue;
+    }
+
+    start.mAcc->AppendTextTo(text, start.mOffset,
+                             segment.End().mOffset - start.mOffset);
   }
 
   return nsCocoaUtils::ToNSString(text);
@@ -297,7 +301,7 @@
 
   if ((mRange.Start().mAcc == mRange.End().mAcc) &&
       (mRange.Start().mAcc->ChildCount() == 0) &&
-      (mRange.Start().mAcc->State() & states::EDITABLE)) {
+      (mRange.Start().mAcc->IsTextField())) {
     return str;
   }
 
@@ -313,11 +317,15 @@
   Accessible* runAcc = range.Start().mAcc;
   for (TextLeafRange segment : range) {
     TextLeafPoint start = segment.Start();
+    if (start.mAcc->IsTextField() && start.mAcc->ChildCount() == 0) {
+      continue;
+    }
     TextLeafPoint attributesNext;
     do {
       attributesNext = start.FindTextAttrsStart(eDirNext, false);
       RefPtr<AccAttributes> attributes = start.GetTextAttributes();
-      if (!attributes->Equal(currentRun)) {
+      MOZ_ASSERT(attributes);
+      if (attributes && !attributes->Equal(currentRun)) {
         AppendTextToAttributedString(str, runAcc, text, currentRun);
         text.Truncate();
         currentRun = attributes;